### PR TITLE
Tester Refactor

### PR DESCRIPTION
**Pull request by rsetaluri**
Started refactoring the tester and targets. Wanted to get some feedback on this -- still more to do if the direction is good. The commit msg should describe the new way of doing things.

@leonardt what do you think?

### CODE REVIEW

File: tests/test_tester.py [hunk 1]
```diff
@@ -10,9 +10,9 @@ def test_tester_basic():
     tester = fault.Tester(circ)
```
Comment (rsetaluri) [inline tests/test_tester.py]: the serialization tests should be disjoint from the tester tests. the tester tests should just ensure that the series of actions generated is accurate (mostly trivial).

File: tests/test_value_utils.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+
+class Foo(m.Circuit):
+    IO = ["a", m.In(m.Bit),
+          "b", m.In(m.Bits(8)),
+          "c", m.In(m.Array(12, m.Bit)),
+          "d", m.In(m.Array(16, m.Array(20, m.Bit)))]
+
+
+def test_all():
```
Comment (rsetaluri) [inline tests/test_value_utils.py]: This test could be structured better, but just wanted to get in all the test functionality so we're sure nothing is broken.

File: fault/verilator_target.py [hunk 3]
```diff
@@ -1,219 +1,99 @@
...
+    def generate_action_code(i, action):
+        if isinstance(action, actions.Poke):
+            return [f"top->{action.port.name} = {action.value};"]
+        if isinstance(action,actions.Expect):
+            return [f"my_assert(top->{action.port.name}, "
+                    f"{action.value}, "
+                    f"{i}, \"{action.port.name}\");"]
+        if isinstance(action, actions.Eval):
+            return ["top->eval();"]
+        if isinstance(action, actions.Step):
```
Comment (rsetaluri) [inline fault/verilator_target.py]: TODO

File: fault/tester.py [hunk 4]
```diff
@@ -1,146 +1,85 @@
...
+
+    def __eval(self):
+        self.vectors.append(self.vectors[-1].copy())
+        for port in self.circuit.interface.ports.values():
+            if port.isinput():
+                index = self.port_to_index[port]
+                self.vectors[-1][index] = make_value(port, AnyValue)
+
+    def process(self, action):
+        if isinstance(action, (actions.Poke, actions.Expect)):
```
Comment (rsetaluri) [inline fault/tester.py]: need to deal with nested arrays here (i.e. `tester.poke(circ.I[2])`)

File: fault/verilator_target.py [hunk 5]
```diff
@@ -1,219 +1,99 @@
...
-        self._directory = directory
-        self._flags = flags
+    def __init__(self, circuit, actions, directory="build/", flags=[]):
+        super().__init__(circuit, actions)
+        self.directory = Path(directory)
+        self.flags = flags
+
+    @staticmethod
+    def generate_action_code(i, action):
+        if isinstance(action, actions.Poke):
```
Comment (rsetaluri) [inline fault/verilator_target.py]: need to deal with nested arrays --> flatten names

Comment (leonardt) [reply fault/verilator_target.py]: Yes, the array flattening stuff is quite annoying, I'm not sure if there's a less complex solution to this.

File: tests/test_value_utils.py (hunk 2 above)
Comment (leonardt) [reply tests/test_value_utils.py]: pytest parametrization might be useful here (see https://docs.pytest.org/en/latest/parametrize.html)

File: tests/test_tester.py (hunk 1 above)
Comment (rsetaluri) [reply tests/test_tester.py]: Done,

File: fault/verilator_target.py (hunk 3 above)
Comment (rsetaluri) [reply fault/verilator_target.py]: Done.

File: fault/tester.py (hunk 4 above)
Comment (rsetaluri) [reply fault/tester.py]: Done.

File: fault/verilator_target.py (hunk 5 above)
Comment (rsetaluri) [reply fault/verilator_target.py]: Done.

**Comment by rsetaluri** [comment]
@leonardt this should be ready for final review.

**Comment by leonardt** [comment]
@rsetaluri looks like the most recent checks passed. Is this ready for a final review?

**Comment by rsetaluri** [comment]
@leonardt yes it is ready for final review

File: tests/test_magma_simulator_target.py [hunk 6]
```diff
@@ -1,51 +1,65 @@
...
 
 
-def test_python_simulator_target_nested_arrays(backend):
+def test_magma_simulator_target_nested_arrays(backend):
+    def _BV(val):
+        return BitVector(val, num_bits=4)
     circ = common.TestNestedArraysCircuit
-    tester = fault.Tester(circ)
-    expected = [random.randint(0, (1 << 4) - 1) for i in range(3)]
+    expected = [_BV(random.randint(0, (1 << 4) - 1)) for i in range(3)]
```
Comment (leonardt) [inline tests/test_magma_simulator_target.py]: This could use `random.rand_bv`

**Pull request merged**